geth/transactions/txqueue_manager_test.go

```diff
@@ -181,6 +172,59 @@ func (s *TxQueueTestSuite) TestCompleteTransactionMultipleTimes() {
 	s.Equal(txCount-1, inprogressTx, "txs expected to be reported as inprogress")
 }
 
+// testTxSender implements ethereum.TransactionSender
+// for testing purposes.
+type testTxSender struct {
+	tx  *common.QueuedTx
+	s   *TxQueueTestSuite
+	acc *common.SelectedExtKey
+}
+
+// SendTransaction is redefined here to inject into
+// CompleteTransaction execution via Manager. Checks
+// transaction values.
+func (sender *testTxSender) SendTransaction(ctx context.Context, tx *types.Transaction) error {
+	s := sender.s
+	s.Equal(sender.tx.Args.Gas.ToInt(), tx.Gas())
+	s.Equal(sender.tx.Args.GasPrice.ToInt(), tx.GasPrice())
+	// Sign the transaction externally.
+	config, err := sender.s.nodeManagerMock.NodeConfig()
+	s.NoError(err)
+	chainID := big.NewInt(int64(config.NetworkID))
+	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(chainID), sender.acc.AccountKey.PrivateKey)
+	s.NoError(err)
+	// Verify signature.
+	v1, r1, s1 := tx.RawSignatureValues()
+	v2, r2, s2 := signedTx.RawSignatureValues()
+	s.Equal(v1, v2)
+	s.Equal(r1, r2)
+	s.Equal(s1, s2)
+	return nil
+}
+
+// TestCompleteTransactionValidity tests if CompleteTransaction
+// uses gas and gas price correctly and sends the expected transaction.
+func (s *TxQueueTestSuite) TestCompleteTransactionValidity() {
```

Comment:
please add a test that is called TestCompleteWithUserDefineGasAndGasPrice

Reply:
Sure but I need to define another test struct which implements `ethereum.TransactionSender`.